### ISSUE TITLE
Extend support for JWT to subscribe to multiple channels

### ISSUE THREAD

**Issue by jamesmstone** [diogob/postgres-websockets#58]
Hi @diogob 

Firstly, thanks this is a really cool idea.

I was wondering if it would be possible to extend support for the JWT to support multiple channels.

eg if you have a JWT like this:

Payload:
```
{
  "mode": "rw",
  "channels": ["chat", "location"]
}
```

JWT:  `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJtb2RlIjoicnciLCJjaGFubmVscyI6WyJjaGF0IiwibG9jYXRpb24iXX0.7y8wL8Vcg6Gsr2yHjJ5H9RnK8DL6U5KP9eChCDVSe2k`

A client could the subscribe to all channels allowed by their token by requesting: 
`ws://eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJtb2RlIjoicnciLCJjaGFubmVscyI6WyJjaGF0IiwibG9jYXRpb24iXX0.7y8wL8Vcg6Gsr2yHjJ5H9RnK8DL6U5KP9eChCDVSe2k
`
Or specific channels from their token:

`ws://chat/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJtb2RlIjoicnciLCJjaGFubmVscyI6WyJjaGF0IiwibG9jYXRpb24iXX0.7y8wL8Vcg6Gsr2yHjJ5H9RnK8DL6U5KP9eChCDVSe2k`

but not channels excluded from their token.

eg 
`ws://updates/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJtb2RlIjoicnciLCJjaGFubmVscyI6WyJjaGF0IiwibG9jYXRpb24iXX0.7y8wL8Vcg6Gsr2yHjJ5H9RnK8DL6U5KP9eChCDVSe2k`

would throw an error.

What do you think?

**Comment by diogob** [comment]
It's an interesting feature. Could you share how are you going to use that? Just so I have a better idea of the target use case and what sort of application is being built using this project.

**Comment by jamesmstone** [comment]
Sure, this really stems from a permissions point of view.  Depending on a users permissions they would subscribe to different channels.

Imagine a db that has different database users for different roles. 

eg a `admin`  users,  `approved` user `anonymous` users etc. 

There are some tables in the system depending on your role you only have permission to view certain table columns.

For example

```
create schema data;
create table data.todo (
	id    serial primary key,
	todo  text not null
        author_email text,
);

grant select, insert, update, delete on data.todo to user;

-- anonymous users can only request specific columns from this view
grant select (id, todo) on data.todo to anonymous;

CREATE OR REPLACE FUNCTION notify_todo()
    RETURNS TRIGGER AS
$$
BEGIN
    PERFORM  pg_notify(
               'postgres-websockets-listener',
               json_build_object('channel', 'approved-todo', 'payload', new)::text
           );
  PERFORM  pg_notify(
               'postgres-websockets-listener',
               json_build_object('channel', 'annonymous-todo', 'payload', new.todo)::text
           );

END;
$$ language 'plpgsql';

CREATE TRIGGER notify_todo
    AFTER UPDATE
    ON data.todo
    FOR EACH ROW
EXECUTE PROCEDURE notify_todo();

```

**Comment by diogob** [comment]
I can't commit to an ETA, but I like the feature and it seems to add little additional complexity. So I should take a stab at it in the next couple of weeks.

**Comment by jamesmstone** [comment]
> I can't commit to an ETA, but I like the feature and it seems to add little additional complexity. So I should take a stab at it in the next couple of weeks.

Would it be as simple as 
1. Updating `type ConnectionInfo` to support a list of channels? (that would be used instead of the current `ByteString` single channel?) 
2. Change `validateClaims` to return the new list of channels
3. Update the `ClaimsSpec` tests 
4. Update `onMessage` somehow... ( this is where I get a bit lost)

**Comment by diogob** [comment]
Yes, you are correct on the 4 points. Add to that change the call to `notifySession`. Both `onMessage` and `notifySession` would have to be mapped over the list of channels.

I'm considering creating the new claims as `channels` and leaving the old `channel` as a deprecated claim for backwards compatibility (if used the claim channels would be ignored).

In case you want to give this PR a shot, the code for the `onMessage` call would look like:
```haskell
flip (onMessage multi) (WS.sendTextData conn . B.payload) <$> channels
```
where channels is a `[ByteString]`

**Comment by jamesmstone** [comment]
By this:
>I'm considering creating the new claims as channels and leaving the old channel as a deprecated claim for backwards compatibility (if used the claim channels would be ignored).

Do you mean treating a `channel:"a"` as equivalent to providing `channels: ["a"`]


>In case you want to give this PR a shot, the code for the onMessage call would look like:
> ```
> flip (onMessage multi) (WS.sendTextData conn . B.payload) <$> channels
> ```
> 

Ohh gee, I am not really a Haskell programmer at all, and it sounds like you have a pretty good idea on what is needed. Happy though to give a PR a go. But would love to maybe have a quick video call to discuss with you first.

**Comment by diogob** [comment]
> Do you mean treating a `channel:"a"` as equivalent to providing `channels: ["a"`]

That is correct

> Ohh gee, I am not really a Haskell programmer at all, and it sounds like you have a pretty good idea on what is needed. Happy though to give a PR a go. But would love to maybe have a quick video call to discuss with you first.

Sure, I've messaged you on Keybase.

**Comment by jamesmstone** [comment]
Thanks, 

This is going to sound terrible. I have forgotten my keybase credentials. I have initiated a reset (7 days remaining)  

Is there any other way you can think of communicating over

**Comment by jamesmstone** [comment]
I just tried sending you a message on [Gitter](https://gitter.im/)

**Comment by diogob** [comment]
I'm closing this since support for multiple channels got merged into the 0.8.0.0 release.
